x/escrow: change the timeout precision to seconds

husio opened the pull request:
A new weave.UnixTime type was added that represents a time as POSIX
time. This is a convinience type used to cast int64 protobuf value to a
more meaningful one.

resolve #447

webmaster128 commented on time.go:
```diff
@@ -0,0 +1,81 @@
+
+// UnmarshalJSON supports unmarshaling both as time.Time and from a number.
+// Usually a number is used as a representation of this time in JSON but it is
+// convinient to use a string format in configurations (ie genesis file).
```
I think we should document how fractional seconds are handled in this case

webmaster128 commented on time.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// Add modifies this UNIX time by given duration. This is compatible with
+// time.Time.Add method.
```
I think we should document how fractional seconds are handled in this case

webmaster128 commented on time.go:
```diff
@@ -0,0 +1,81 @@
+	return t + UnixTime(d/time.Second)
+}
+
+// AsUnixTime converts given Time structure into its UNIX time representation.
```
I think we should document how fractional seconds are handled in this case

webmaster128 commented on time.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// IsZero returns true if this time represents a zero value.
+func (t UnixTime) IsZero() bool {
```
Is this used?

webmaster128 commented on time.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// Validate returns an error if this time value is invalid.
+func (t UnixTime) Validate() error {
```
We could potentially allow negative values as well. Many implementations allow them.

One useful limitation would be the safe integer range for compatibility with JS/JSON numbers: [[MIN_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MIN_SAFE_INTEGER), [MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER)]

webmaster128 commented on x/escrow/msg_test.go:
```diff
@@ -34,7 +34,7 @@ func TestCreateEscrowMsg(t *testing.T) {
 	d := weave.Condition("foobar")
 
-	timeout := time.Now()
+	timeout := weave.AsUnixTime(time.Now())
```
A native `Now()` on the new type would be nice

husio replied:
I am not sure what you are suggesting. There are no constructors in Go. I could add `weave.UnixTimeNow()` function, but I think having smaller API is in this case better.

webmaster128 replied:
Ah, so only instance methods are possible (like .IsZero(), .Validate()) in UnixTime but no factory functions, right? I see, yeah, not so important. Smaller API is good.

webmaster128 commented on x/escrow/time.go:
```diff
@@ -12,10 +10,10 @@ import (
 		panic("block time is not present")
 	}
-	return t.Before(blockNow)
+	return t.Time().Before(blockNow)
```
This comparison should happen on `weave.UnixTime`, not on `Time`.

E.g. `100.1 before 100.2` is true on Time but `UnixTime(100.1) < UnixTime(100.2)` is false

husio commented:
@webmaster128 I have addressed all comments in the last commit https://github.com/iov-one/weave/pull/490/commits/3fca5d5d69aef5733318d0e5fad897467ea08fd4 Please have a look once more.

webmaster128 commented on time.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+// UNIX time value of 0001-01-01T00:00:00Z
+const zeroUnixTime = -62135596800
```
This is indeed the lowest value that Python supports:

```
>>> datetime.datetime.utcfromtimestamp(-62135596800)
datetime.datetime(1, 1, 1, 0, 0)
```

and much stricter than Number.MIN_SAFE_INTEGER. Makes sense to me for the general time type

husio replied:
I call it zero time, because I got this value by extracting unix time from zero `time.Time` value. Go's `Time` does not seem to support anything that is before that point in time.

webmaster128 commented on time.go:
```diff
@@ -0,0 +1,85 @@
+func (t UnixTime) Validate() error {
+	if t < zeroUnixTime {
+		return errors.Wrap(errors.ErrInvalidState, "time must be an A.D. value")
+	}
```
We should have an upper limit as well to ensure each values is <= Number.MAX_SAFE_INTEGER = 9007199254740991.

What about 9999-12-31T23:59:59Z?

```
>>> datetime.datetime.utcfromtimestamp(253402300799)
datetime.datetime(9999, 12, 31, 23, 59, 59)
```

husio replied:
Go supports way higher value but I don't think we will ever need it https://play.golang.org/p/5zD6pPw73RS

9007199254740991 is also good enough for the max value. I think year 285428751 is not our problem anymore :trollface: 
https://play.golang.org/p/dcq0-78mW67

webmaster128 replied:
Then let's use 9999-12-31T23:59:59Z for some kind of symmetry. 0001-01-01T00:00:00Z to 9999-12-31T23:59:59Z is good to document and is in the safe integer range of JS/JSON.

ethanfrey commented on time.go:
```diff
@@ -0,0 +1,85 @@
+		return nil
+	}
+
+	var stdtime time.Time
```
I like this as fallback for parsing genesis. Nice touch.

ethanfrey commented on time.go:
```diff
@@ -0,0 +1,85 @@
+	var n int64
+	if err := json.Unmarshal(raw, &n); err == nil {
+		unix := UnixTime(n)
+		if err := unix.Validate(); err != nil {
```
:+1:

ethanfrey commented on time.go:
```diff
@@ -0,0 +1,85 @@
+
+// String returns the usual string representation of this time as the time.Time
+// structure would.
+func (t UnixTime) String() string {
```
:+1: for usability
